### PR TITLE
Record moe routing decisions during inference. 

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
# What does this PR do ?
Enabled router recording during inference of MoEs, using the pre-existing `--moe-enable-routing-replay` argument. This can be used i. to study routing decisions of inference requests and ii. router replay in RL, which is a popular technique to reduce logprob mismatches.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/moe/router.py
Comment (tdene) [inline megatron/core/transformer/moe/router.py:1]: Can you remove the diff from this file?

File: megatron/core/transformer/moe/router_replay.py [hunk 1]
```diff
@@ -159,3 +183,29 @@ def get_replay_topk(
...
+        Args:
+            buffer: Tensor of shape [max_tokens, topk] to copy routing indices into.
+        """
+        self.static_buffer = buffer
+
+    def clear_static_buffer(self):
+        """Clears the static buffer."""
+        self.static_buffer = None
+
+    def record_indices(self, topk_indices: torch.Tensor):
```
Comment (yanring) [inline megatron/core/transformer/moe/router_replay.py:195]: Duplicate `record_indices` method definition with L122

File: tests/functional_tests/test_cases/moe/gpt_dynamic_inference_tp4_etp1_pp1_ep8_16B_logitsmatch_cudagraph_zmq/model_config.yaml [hunk 2]
```diff
@@ -84,6 +84,7 @@ MODEL_ARGS:
   --inference-dynamic-batching-max-requests: 512
   --inference-logging-step-interval: 1
   --sequence-parallel: true
+  --moe-enable-routing-replay: true
```
Comment (wdykas) [inline tests/functional_tests/test_cases/moe/gpt_dynamic_inference_tp4_etp1_pp1_ep8_16B_logitsmatch_cudagraph_zmq/model_config.yaml:87]: What is actually being tested by adding this to the functional test? Just that it won't crash?

File: megatron/core/inference/contexts/routing_metadata.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        self._ensure_buffer_allocated()
+        if self.routing_indices_buffer is not None:
+            RouterReplay.set_global_static_buffers(self.routing_indices_buffer)
+
+    def disable_static_buffer_recording(self) -> None:
+        """Disable static buffer recording, reverting to normal tensor assignment."""
+        RouterReplay.clear_global_static_buffers()
+
+    def reset(self) -> None:
```
Comment (wdykas) [inline megatron/core/inference/contexts/routing_metadata.py]: why do we have this function if all it does is pass?

**Review by wdykas — APPROVED** [review]

**Comment by wdykas** [comment]
Does this work with PP I see that there is an argument for number of layers but not sure where the number of layers is coming from?

File: tests/functional_tests/test_cases/moe/gpt_dynamic_inference_tp4_etp1_pp1_ep8_16B_logitsmatch_cudagraph_zmq/model_config.yaml (hunk 2 above)
Comment (sidsingh-nvidia) [reply tests/functional_tests/test_cases/moe/gpt_dynamic_inference_tp4_etp1_pp1_ep8_16B_logitsmatch_cudagraph_zmq/model_config.yaml:87]: Yeah for now, just that it will not crash. I should probably add a more stringent test though.

File: megatron/core/transformer/moe/router_replay.py (hunk 1 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/router_replay.py:195]: Thanks for the catch!

**Comment by sidsingh-nvidia** [comment]
/ok to test [452acb2](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/452acb20c2316546de2ad2cfe2363100ef48b75c)

**Comment by sidsingh-nvidia** [comment]
@wdykas, good point. The router replay module maintains a global record of every top-k router on the GPU. And I am grabbing the number of layers from that global record itself (instead of doing math on args.num_layers and what not). So as long as the router replay module registers all topk router layers, we should be good with PP.

File: tests/functional_tests/test_cases/moe/gpt_dynamic_inference_tp4_etp1_pp1_ep8_16B_logitsmatch_cudagraph_zmq/model_config.yaml (hunk 2 above)
Comment (sidsingh-nvidia) [reply tests/functional_tests/test_cases/moe/gpt_dynamic_inference_tp4_etp1_pp1_ep8_16B_logitsmatch_cudagraph_zmq/model_config.yaml:87]: added golden values for router replay

**Comment by sidsingh-nvidia** [comment]
@jon-barker I have a functional test now that checks the router record output against some golden values.

**Comment by sidsingh-nvidia** [comment]
/ok to test [601b77d](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/601b77d17222b3e953fe8a922abe6a00406aaff6)

**Review by tdene — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Comment by sidsingh-nvidia** [comment]
/ok to test [045b07b](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/045b07bd6a50c862f85487a993a43cb683bd4ea2)

**Comment by sidsingh-nvidia** [comment]
/ok to test [48bb9b9](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/48bb9b9628af7c8059103ef66d4ef7848d66b5cc)

**Comment by sidsingh-nvidia** [comment]
/ok to test [b632151](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/b632151e9f75dc8dd635c6f5ba95f21de4628651)

**Comment by sidsingh-nvidia** [comment]
/ok to test [e3c6cab](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/e3c6cab999f17e8880f93bbe8bdd2dc0330e29ee)

**Comment by sidsingh-nvidia** [comment]
/ok to test [061228c](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/061228c72b11c042598ab91118d5ad0e243e8233)

**Comment by sidsingh-nvidia** [comment]
/ok to test [e084310](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/e084310ed64c4d99b4cb88a63a1558a0f83fdc68)

**Comment by sidsingh-nvidia** [comment]
/ok to test [c05da3b](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/c05da3b839bc96c43175b999c324e2abe2c4d775)

**Comment by sidsingh-nvidia** [comment]
/ok to test [a99ddee](https://github.com/NVIDIA/Megatron-LM/pull/3034/commits/a99ddee93865eaeb752b57d5839a1c7c65d8be34)

**Comment by ko3n1g** [comment]
🔄 Merge queue validation started!

You can track the progress here: https://github.com/NVIDIA/Megatron-LM/actions/runs/21924549756

**Comment by ko3n1g** [comment]
🔄 Merge queue validation started!

You can track the progress here: https://github.com/NVIDIA/Megatron-LM/actions/runs/21927646884

**Comment by ko3n1g** [comment]
🔄 Merge queue validation started!

You can track the progress here: https://github.com/NVIDIA/Megatron-LM/actions/runs/21928298430
